File: C_resources/quick_sort.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>

void quicksort(int array[], int start, int end)
{
    int i, j;
    int pilot;
    int tmp;

    if (start >= end) return;

    pilot = array[start];
    i = start+1;
    j = end;

    while (i < j) {
        while (array[i] < pilot) {
            i++;
        }

        while (array[j] > pilot) {
            j--;
        }

        if (i < j) {
            tmp = array[i];
            array[i] = array[j];
            array[j] = tmp;
        }
    }

    tmp = array[start];
    array[start] = array[j];
    array[j] = tmp;

    quicksort(array, start, j-1);
    quicksort(array, j+1, end);
}

int main() {
    int n;
    char string[100];
    int array[10];
    char* ptr;
    int i = 0;

    printf("Please input the record num: ");
    scanf("%d", &n);
    getchar();

    gets(string);

    ptr = strtok(string, " ");
    array[i++] = atoi(ptr);
    while (i < n) {
        ptr = strtok(NULL, " ");
        array[i++] = atoi(ptr);
    }

    quicksort(array, 0, n-1);

    printf("n=%d\n", n);
    for (i=0; i<n; i++) {
        printf("array[%d]=%d\n", i, array[i]);
    }

    return 0;
}

File: Python_resources/misc/quicksort.py
def quicksort(A, start, end):
    if (start >= end):
        return
    pivot = A[start]
    low = start
    high = end
    while (low < high):
        while (low < high and A[high] >= pivot):
            high -= 1
        A[low] = A[high]
        while (low < high and A[low] <= pivot):
            low += 1
        A[high] = A[low]
    A[low] = pivot
    quicksort(A, start, low-1)
    quicksort(A, low+1, end)

A = [1,4,9,3,4,1,5,2,8]
#A = [2,1,3]
quicksort(A, 0, len(A)-1)
print A
